### PR TITLE
Proto conversion for deadletter

### PR DESCRIPTION
**Pull request by mkcode92**

### CODE REVIEW

**Comment by torbsto** [comment]
You have to include the protoc artifact in build file of the proto project, see [brute-force-serde](https://github.com/bakdata/kafka-brute-force-serde/blob/main/brute-force-serde/build.gradle.kts):
```kotlin
import com.google.protobuf.gradle.protobuf
import com.google.protobuf.gradle.protoc

...

protobuf {
    protoc {
        artifact = "com.google.protobuf:protoc:3.18.1"
    }
}
```

**Comment by torbsto** [comment]
You can increase Gradle's heap space in `gradle.properties`, e.g. `org.gradle.jvmargs=-Xmx2048m`

File: avro/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+description = "A library for error handling in Kafka Streams."
```
Comment (philipp94831) [inline avro/build.gradle.kts:1]: Please update the description

File: avro/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    main {
+        resources {
+            srcDirs("src/main/avro")
+        }
+    }
+}
+
+dependencies {
+    val kafkaVersion: String by project
+    api(group = "org.apache.kafka", name = "kafka-streams", version = kafkaVersion)
```
Comment (philipp94831) [inline avro/build.gradle.kts:19]: Please only use the dependencies that are needed for this module

File: avro/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    val avroVersion: String by project
+    api(group = "org.apache.avro", name = "avro", version = avroVersion)
+    implementation(project(":core"))
```
Comment (philipp94831) [inline avro/build.gradle.kts:22]: should be implementation

File: avro/src/main/java/com/bakdata/kafka/AvroDeadLetterConverter.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.bakdata.kafka;
+
+/**
+ * Convert a {@code DeadLetterDescription} to an avro {@code DeadLetter}
```
Comment (philipp94831) [inline avro/src/main/java/com/bakdata/kafka/AvroDeadLetterConverter.java:28]: Avro

File: avro/src/main/java/com/bakdata/kafka/AvroDeadLetterTransformer.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * SOFTWARE.
+ */
+
+package com.bakdata.kafka;
+
+import lombok.NonNull;
+import org.apache.kafka.streams.kstream.ValueTransformerSupplier;
+
+/**
+ * Transform error representations to an avro {@code DeadLetter}
```
Comment (philipp94831) [inline avro/src/main/java/com/bakdata/kafka/AvroDeadLetterTransformer.java:31]: Avro

File: avro/src/main/java/com/bakdata/kafka/AvroDeadLetterTransformer.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * @param <V> type of the input value
+ */
+public final class AvroDeadLetterTransformer<V> extends DeadLetterTransformer<V, DeadLetter> {
```
Comment (philipp94831) [inline avro/src/main/java/com/bakdata/kafka/AvroDeadLetterTransformer.java:35]: why do we need this class? can't the use just supply the converter?

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(SoftAssertionsExtension.class)
+class AvroDeadLetterConverterTest {
+
+    static final AvroDeadLetterConverter converter = new AvroDeadLetterConverter();
+
+    @Test
+    void shouldConvertDeadletterDescriptionWithOptionalFields(final SoftAssertions softly) {
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java:38]: I know the old tests do it like this but the new way is to have a field with type `SoftAssertions` and annotate it with `@InjectSoftAssertions`. You can then remove the param

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                        .build())
+                .description("description")
+                .topic("topic")
+                .partition(1)
+                .offset(1L)
+                .build();
+
+        final DeadLetter deadLetter = converter.convert(deadLetterDescription);
+        softly.assertThat(deadLetter.getInputValue()).isPresent();
+        softly.assertThat(deadLetter.getInputValue().get()).isEqualTo("inputValue");
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java:54]: Do not use `.get()`. Do `softly.assertThat(deadLetter.getInputValue()).hasValue("inputValue");` instead. Thereby you can also remove the isPresent assertion

File: core/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    testFixturesImplementation(
+        group = "com.bakdata.fluent-kafka-streams-tests",
+        name = "fluent-kafka-streams-tests-junit5",
+        version = "2.4.2"
+    )
+    testFixturesImplementation(group = "org.apache.logging.log4j", name = "log4j-slf4j-impl", version = log4jVersion)
+    testFixturesImplementation(group = "io.confluent", name = "kafka-streams-avro-serde", version = confluentVersion)
+}
+
+avro {
```
Comment (philipp94831) [inline core/build.gradle.kts:55]: Is avro still needed in this module?

File: core/src/test/java/com/bakdata/kafka/TestDeadLetterSerde.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    private static final Serializer<DeadLetterDescription> serializer = (topic, data) -> {
+        try {
+            return objectMapper.writeValueAsBytes(data);
+        } catch (final JsonProcessingException e) {
+            throw new SerializationException(e);
+        }
+    };
+    private static final Deserializer<DeadLetterDescription> deserializer = (topic, data) -> {
+        try {
+            final JsonNode json = objectMapper.readTree(data);
```
Comment (philipp94831) [inline core/src/test/java/com/bakdata/kafka/TestDeadLetterSerde.java:47]: why don't you do `.readValue(data, DeadLetterDescription.class)`?

File: core/src/test/java/com/bakdata/kafka/TestDeadLetterTransformer.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.bakdata.kafka;
+
+import lombok.NonNull;
+import org.apache.kafka.streams.kstream.ValueTransformerSupplier;
+
+public final class TestDeadLetterTransformer<V> extends DeadLetterTransformer<V, DeadLetterDescription> {
```
Comment (philipp94831) [inline core/src/test/java/com/bakdata/kafka/TestDeadLetterTransformer.java:30]: same question as above

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: core/build.gradle.kts (hunk 9 above)
Comment (mkcode92) [reply core/build.gradle.kts:55]: For the TestValue used in the ErrorUtilTest, as ErrorUtil supports custom Avro conversion.

File: avro/build.gradle.kts (hunk 3 above)
Comment (mkcode92) [reply avro/build.gradle.kts:20]: guess you mean L21?

Comment (philipp94831) [reply avro/build.gradle.kts:20]: Ah^^ Line was correct but I messed up: Should be api

File: settings.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+pluginManagement {
+    repositories {
+        gradlePluginPortal()
+    }
+}
+
+rootProject.name = "error-handling"
+include("core")
+include("avro")
+include("proto")
```
Comment (torbsto) [inline settings.gradle.kts:10]: In large-message and brute-force, we prefix the modules with the corresponding project. So in this case:
- `error-handling-core`
- `error-handling-avro`
- `error-handling-proto`

Otherwise, we would publish an artifact called `com.bakdata.kafka:avro`, which is too generic I think.

**Review by torbsto — COMMENTED** [review]
Could you also update the ReadMe?

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                        .build())
+                .description("description")
+                .topic("topic")
+                .partition(1)
+                .offset(1L)
+                .build();
+
+        final DeadLetter deadLetter = converter.convert(deadLetterDescription);
+        this.softly.assertThat(deadLetter.getInputValue()).isPresent();
+        this.softly.assertThat(deadLetter.getInputValue().get()).isEqualTo("inputValue");
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java:57]: please use proper optional assertions here

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+@ExtendWith(MockitoExtension.class)
+@ExtendWith(SoftAssertionsExtension.class)
+class AvroDeadLetterTransformerTest extends ErrorCaptureTopologyTest {
+    @InjectSoftAssertions
+    private SoftAssertions softly;
+    private static final String ERROR_TOPIC = "errors";
+    private static final String OUTPUT_TOPIC = "output";
+    private static final String INPUT_TOPIC = "input";
+    private static final Serde<String> STRING_SERDE = Serdes.String();
+    public static final String DEADLETTER_DESCRIPTION = "Description";
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:59]: DEAD_LETTER_DESCRIPTION or just DESCRIPTION

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Override
+    protected void buildTopology(final StreamsBuilder builder) {
+        final KStream<Integer, String> input = builder.stream(INPUT_TOPIC, Consumed.with(null, STRING_SERDE));
+        final KStream<Integer, ProcessedKeyValue<Integer, String, String>> mapped =
+                input.map(ErrorCapturingKeyValueMapper.captureErrors(this.mapper));
+        mapped.flatMapValues(ProcessedKeyValue::getValues)
+                .to(OUTPUT_TOPIC, Produced.valueSerde(STRING_SERDE));
+        mapped.flatMap(ProcessedKeyValue::getErrors)
+                .transformValues(DeadLetterTransformer.create(DEADLETTER_DESCRIPTION, new AvroDeadLetterConverter()))
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:73]: @torbsto What do you think about making this a bit more readable with sth like `DeadLetterTransformer.transform(toAvro(), DEADLETTER_DESCRIPTION)` where toAvro is just a static method in AvroDeadLetterConverter? Even more advanced would be sth like `describedAs(DESCRIPTION).transformedTo(avro())`

File: avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+                .to(ERROR_TOPIC);
+    }
+
+    @Test
+    void shouldConvertAndSerializeProtoDeadLetter() {
```
Comment (philipp94831) [inline avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:78]: the name of the test seems wrong

File: avro/build.gradle.kts [hunk 17]
```diff
@@ -0,0 +1,42 @@
+description = "Serializing errors in Avro."
```
Comment (philipp94831) [inline avro/build.gradle.kts:1]: Transform dead letters in Kafka Streams applications to Avro format

File: core/build.gradle.kts [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+            srcDirs("src/test/avro")
+        }
+    }
+}
+
+dependencies {
+    val kafkaVersion: String by project
+    api(group = "org.apache.kafka", name = "kafka-streams", version = kafkaVersion)
+    val avroVersion: String by project
+    api(group = "org.apache.avro", name = "avro", version = avroVersion)
```
Comment (philipp94831) [inline core/build.gradle.kts:19]: this can be implementation as it is only used internally in the ErrorUtil but is not part of any method signature

File: avro/build.gradle.kts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+dependencies {
+    val avroVersion: String by project
+    api(group = "org.apache.avro", name = "avro", version = avroVersion)
+    api(project(":core"))
+
+    val junitVersion: String by project
+    testRuntimeOnly(group = "org.junit.jupiter", name = "junit-jupiter-engine", version = junitVersion)
+    testImplementation(testFixtures(project(":core")))
+    testImplementation(group = "org.jooq", name = "jool", version = "0.9.14")
+    val mockitoVersion = "3.12.4"
```
Comment (philipp94831) [inline avro/build.gradle.kts:26]: please move to gradle.properties as we reuse it in multiple modules

File: core/src/main/java/com/bakdata/kafka/DeadLetterDescription.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    @Value
+    @Jacksonized
+    public static class Cause {
+        String message;
+        String stackTrace;
+        String errorClass;
+    }
+
+    @NonNull String description;
+    Cause cause;
```
Comment (philipp94831) [inline core/src/main/java/com/bakdata/kafka/DeadLetterDescription.java:53]: I think this can be NonNull. Please compare it to the original avro model

File: core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * {@link ValueTransformer} that creates a {@code DeadLetter} from a processing error.
+ *
+ * @param <V> type of value
+ * @param <VR> the DeadLetter type
+ */
+@Getter
+@RequiredArgsConstructor
+public class DeadLetterTransformer<V, VR> implements ValueTransformer<ProcessingError<V>, VR> {
```
Comment (philipp94831) [inline core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java:44]: I would rename VR to T as it is unrelated to V and VR usually means "value returned"

File: core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+     * Transforms captured errors for serialization
+     *
+     * <pre>{@code
+     * // Example, this works for all error capturing topologies
+     * final KeyValueMapper<K, V, KeyValue<KR, VR>> mapper = ...;
+     * final KStream<K, V> input = ...;
+     * final KStream<KR, ProcessedKeyValue<K, V, VR>> processed = input.map(captureErrors(mapper));
+     * final KStream<KR, VR> output = processed.flatMapValues(ProcessedKeyValue::getValues);
+     * final KStream<K, ProcessingError<V>> errors = processed.flatMap(ProcessedKeyValue::getErrors);
+     * errors.transformValues(DeadLetterTransformer.create("Description", new AvroDeadLetterConverter()))
```
Comment (philipp94831) [inline core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java:59]: Please extract the converter as a variable with `= ...` as Avro is not part of this module

File: core/src/test/java/com/bakdata/kafka/ProcessingErrorTest.java [hunk 23]
```diff
@@ -85,10 +57,4 @@ void shouldAllowNullValue() {
                 .isNotNull();
     }
 
-    @Test
-    void shouldNotAllowNullDescription() {
```
Comment (philipp94831) [inline core/src/test/java/com/bakdata/kafka/ProcessingErrorTest.java:89]: This test is missing in the avro module I think. Although it can also be part of this module as the description of DeadLetterDescription should be non nullable

File: core/src/test/java/com/bakdata/kafka/TestDeadLetterSerde.java [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    };
+    private static final Deserializer<DeadLetterDescription> deserializer = (topic, data) -> {
+        try {
+            return objectMapper.readValue(data, DeadLetterDescription.class);
+        } catch (IOException e) {
+            throw new SerializationException(e);
+        }
+    };
+
+    private static String asTextOrNull(final JsonNode json) {
```
Comment (philipp94831) [inline core/src/test/java/com/bakdata/kafka/TestDeadLetterSerde.java:53]: unused

File: proto/build.gradle.kts [hunk 25]
```diff
@@ -0,0 +1,36 @@
+import com.google.protobuf.gradle.protobuf
+import com.google.protobuf.gradle.protoc
+
+description = "Serializing errors in protobuf."
+
+plugins {
+    idea
```
Comment (philipp94831) [inline proto/build.gradle.kts:7]: Is that plugin really needed?

File: proto/build.gradle.kts [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+    id("com.google.protobuf") version "0.8.18"
+}
+
+dependencies {
+    api(project(":core"))
+    implementation(group = "com.google.protobuf", name = "protobuf-java", version = "3.18.1")
```
Comment (philipp94831) [inline proto/build.gradle.kts:13]: That should probably be api

File: proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto [hunk 27]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
+
+package bakdata.kafka.proto.v1;
```
Comment (philipp94831) [inline proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto:3]: I don't know what is best practice here but why not include com?

File: proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterConverterTest.java [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(SoftAssertionsExtension.class)
+class ProtoDeadLetterConverterTest {
+
+    static final ProtoDeadLetterConverter converter = new ProtoDeadLetterConverter();
+
+    @Test
+    void shouldConvertDeadletterDescriptionWithOptionalFields(final SoftAssertions softly) {
```
Comment (philipp94831) [inline proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterConverterTest.java:39]: please use InjectSoftAssertions

File: proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterTransformerTest.java [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+
+    private Serde<DeadLetter> configuredDeadLetterSerde = null;
+    @Mock
+    KeyValueMapper<Integer, String, KeyValue<Integer, String>> mapper;
+
+    Serde<DeadLetter> deadLetterSerde() {
+        if (this.configuredDeadLetterSerde == null) {
+            this.configuredDeadLetterSerde = new KafkaProtobufSerde<>(DeadLetter.class);
+            this.configuredDeadLetterSerde.configure(Map.of(
+                    AbstractKafkaSchemaSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, "mock://test"), false);
```
Comment (philipp94831) [inline proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterTransformerTest.java:77]: why not use the SR mock?

File: settings.gradle [hunk 30]
```diff
@@ -1,7 +0,0 @@
-pluginManagement {
```
Comment (philipp94831) [inline settings.gradle:1]: why did you delete this file?

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: core/src/test/java/com/bakdata/kafka/ProcessingErrorTest.java (hunk 23 above)
Comment (mkcode92) [reply core/src/test/java/com/bakdata/kafka/ProcessingErrorTest.java]: Introduced some Nullability checks in DeadLetterDescriptionTest...

File: proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto (hunk 27 above)
Comment (mkcode92) [reply proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto]: Following the example here: https://docs.buf.build/best-practices/style-guide - not explicitly stated, however.

File: proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterTransformerTest.java (hunk 29 above)
Comment (mkcode92) [reply proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterTransformerTest.java]: The topology and the SR mock are not yet available when building the topology. Do you see a non-ugly workaround?

Comment (philipp94831) [reply proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterTransformerTest.java]: Well here it seems to work: https://github.com/bakdata/kafka-error-handling/pull/13/files#diff-74f44e129bb101e8fd8c90cb9449e18f7163e6f49a6fb64593639d574ffcf678

**Comment by philipp94831** [comment]
Let's test if the library is usable. Please add `ref: tmp/upload-snapshot` to the azure template https://github.com/bakdata/kafka-error-handling/blob/master/azure-pipelines.yml#L18. It will then be published to Sonatype Snapshot Repo (usually only happens on master branch)

**Comment by mkcode92** [comment]
tested the snapshot manually in one of our projects, works both for avro and proto as expected.

**Comment by mkcode92** [comment]
At which version should we start with the new libraries?

**Comment by philipp94831** [comment]
> At which version should we start with the new libraries?

This will be handled by the release pipeline. We will use 1.3 or 2.0, not sure yet

File: error-handling-avro/build.gradle.kts [hunk 31]
```diff
@@ -0,0 +1,43 @@
+description = "Transform dead letters in Kafka Streams applications to Avro format."
+
+plugins {
+    `java-library`
```
Comment (philipp94831) [inline error-handling-avro/build.gradle.kts:4]: not needed because the parent project applies it to subprojects

File: error-handling-avro/src/main/java/com/bakdata/kafka/AvroDeadLetterConverter.java [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * <pre>{@code
+     * // Example, this works for all error capturing topologies
+     * final KeyValueMapper<K, V, KeyValue<KR, VR>> mapper = ...;
+     * final KStream<K, V> input = ...;
+     * final KStream<KR, ProcessedKeyValue<K, V, VR>> processed = input.map(captureErrors(mapper));
+     * final KStream<KR, VR> output = processed.flatMapValues(ProcessedKeyValue::getValues);
+     * final KStream<K, ProcessingError<V>> errors = processed.flatMap(ProcessedKeyValue::getErrors);
+     * errors.transformValues(AvroDeadLetterConverter.asTransformer("Description"))
+     *       .to(ERROR_TOPIC);
```
Comment (philipp94831) [inline error-handling-avro/src/main/java/com/bakdata/kafka/AvroDeadLetterConverter.java:61]: maybe extract a variable here so that it is clear that you get a `KStream<K, DeadLetter>`

File: error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+import org.assertj.core.api.junit.jupiter.SoftAssertionsExtension;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(SoftAssertionsExtension.class)
+class AvroDeadLetterConverterTest {
+    @InjectSoftAssertions
+    private SoftAssertions softly;
+
+    static final AvroDeadLetterConverter converter = new AvroDeadLetterConverter();
```
Comment (philipp94831) [inline error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterConverterTest.java:38]: Why do we need a field here? A local variable should do the job as well?

File: error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+import org.assertj.core.api.junit.jupiter.SoftAssertionsExtension;
+import org.jooq.lambda.Seq;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+@ExtendWith(SoftAssertionsExtension.class)
+class AvroDeadLetterTransformerTest extends ErrorCaptureTopologyTest {
```
Comment (philipp94831) [inline error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:52]: Please annotate with `@MockitoSettings(strictness = Strictness.STRICT_STUBS)`

File: error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 35]
```diff
@@ -0,0 +1,124 @@
...
+    @InjectSoftAssertions
+    private SoftAssertions softly;
+    private static final String ERROR_TOPIC = "errors";
+    private static final String OUTPUT_TOPIC = "output";
+    private static final String INPUT_TOPIC = "input";
+    private static final Serde<String> STRING_SERDE = Serdes.String();
+    public static final String DEAD_LETTER_DESCRIPTION = "Description";
```
Comment (philipp94831) [inline error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:59]: private (also below)

File: error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+    public static final String ERROR_MESSAGE = "ERROR!";
+
+    @Mock
+    KeyValueMapper<Integer, String, KeyValue<Integer, String>> mapper;
```
Comment (philipp94831) [inline error-handling-avro/src/test/java/com/bakdata/kafka/AvroDeadLetterTransformerTest.java:63]: can be private

File: error-handling-core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java [hunk 37]
```diff
@@ -0,0 +1,101 @@
...
+    public void init(final ProcessorContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public T transform(final ProcessingError<V> error) {
+        final DeadLetterDescription deadLetterDescription = DeadLetterDescription.builder()
+                .inputValue(Optional.ofNullable(error.getValue()).map(ErrorUtil::toString).orElse(null))
+                .cause(DeadLetterDescription.Cause.builder()
+                        .message(error.getThrowable().getMessage())
```
Comment (philipp94831) [inline error-handling-core/src/main/java/com/bakdata/kafka/DeadLetterTransformer.java:85]: extract a variable for throwable

File: error-handling-core/src/test/java/com/bakdata/kafka/DeadLetterDescriptionTest.java [hunk 38]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    void shouldAllowOtherFieldsNull() {
+        assertThat(DeadLetterDescription.builder()
+                .description("foo")
+                .cause(CAUSE)
+                .inputValue(null)
+                .topic(null)
+                .partition(null)
+                .offset(null)
+                .build()).isNotNull();
```
Comment (philipp94831) [inline error-handling-core/src/test/java/com/bakdata/kafka/DeadLetterDescriptionTest.java:34]: shouldn't that more be a `assertThatCode(...).doesNotThrowException()`?

File: error-handling-core/src/test/java/com/bakdata/kafka/ErrorCapturingFlatKeyValueMapperTopologyTest.java [hunk 39]
```diff
@@ -73,8 +74,9 @@ protected void buildTopology(final StreamsBuilder builder) {
         mapped.flatMapValues(ProcessedKeyValue::getValues)
                 .to(OUTPUT_TOPIC, Produced.with(DOUBLE_SERDE, LONG_SERDE));
         mapped.flatMap(ProcessedKeyValue::getErrors)
-                .transformValues(DeadLetterTransformer.createDeadLetter("Description"))
-                .to(ERROR_TOPIC);
+                .transformValues(
+                        DeadLetterTransformer.create("Description", deadLetterDescription -> deadLetterDescription))
+                .to(ERROR_TOPIC, Produced.valueSerde(DEAD_LETTER_SERDE));
```
Comment (philipp94831) [inline error-handling-core/src/test/java/com/bakdata/kafka/ErrorCapturingFlatKeyValueMapperTopologyTest.java:79]: why not configure the default value serde properly?

File: error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto [hunk 40]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package bakdata.kafka.proto.v1;
+
+import "google/protobuf/wrappers.proto";
+
+option java_package = "com.bakdata.kafka.proto.v1";
+option java_multiple_files = true;
+
+message DeadLetter {
```
Comment (philipp94831) [inline error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto:10]: Hm, I'm thinking whether we should use a different name. The package differs from Avro but when we integrate it into the dead-letter-analyzer, we will not be able to import both. What do you think?

File: error-handling-proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterConverterTest.java [hunk 41]
```diff
@@ -0,0 +1,83 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(SoftAssertionsExtension.class)
+class ProtoDeadLetterConverterTest {
+
+    @InjectSoftAssertions
+    private SoftAssertions softly;
+
+    static final ProtoDeadLetterConverter converter = new ProtoDeadLetterConverter();
```
Comment (philipp94831) [inline error-handling-proto/src/test/java/com/bakdata/kafka/ProtoDeadLetterConverterTest.java:40]: see above

File: gradle.properties [hunk 42]
```diff
@@ -1,7 +1,11 @@
...
 org.gradle.caching=true
 org.gradle.parallel=true
+org.gradle.jvmargs=-Xmx2048m
 kafkaVersion=2.8.0
 avroVersion=1.10.2
-confluentVersion=6.2.0
+confluentVersion=7.1.1
 junitVersion=5.7.2
+mockitoVersion=3.12.4
+log4jVersion=2.15.0
```
Comment (philipp94831) [inline gradle.properties:10]: 2.17.2

File: gradle.properties [hunk 43]
```diff
@@ -1,7 +1,11 @@
 version=1.2.6-SNAPSHOT
 org.gradle.caching=true
 org.gradle.parallel=true
+org.gradle.jvmargs=-Xmx2048m
 kafkaVersion=2.8.0
 avroVersion=1.10.2
-confluentVersion=6.2.0
+confluentVersion=7.1.1
```
Comment (philipp94831) [inline gradle.properties:7]: If you need to bump confluent, then please also bump Kafka to 3.1.0. Otherwise leave Kafka, Confluent and fluent Kafka as it was

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto (hunk 40 above)
Comment (mkcode92) [reply error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto:10]: I'll prefix them, also confuses the IDE less.

Comment (philipp94831) [reply error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto:10]: For better backwards compatibility please leave Avro as it was

File: gradle.properties (hunk 43 above)
Comment (mkcode92) [reply gradle.properties:7]: the bump is required for protobuf support in the schemaregistrymock

File: error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto (hunk 40 above)
Comment (philipp94831) [reply error-handling-proto/src/main/proto/bakdata/kafka/proto/v1/deadletter.proto]: @mkcode92 I think you missed that

File: error-handling-proto/build.gradle.kts [hunk 44]
```diff
@@ -0,0 +1,49 @@
...
+protobuf {
+    protoc {
+        artifact = "com.google.protobuf:protoc:3.18.1"
+    }
+}
+
+sourceSets {
+    main {
+        java {
+            srcDir("build/generated/source/proto/main/java")
```
Comment (philipp94831) [inline error-handling-proto/build.gradle.kts:46]: Is that not handled by the proto plugin?

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: error-handling-core/src/test/java/com/bakdata/kafka/ErrorCapturingFlatKeyValueMapperTopologyTest.java (hunk 39 above)
Comment (mkcode92) [reply error-handling-core/src/test/java/com/bakdata/kafka/ErrorCapturingFlatKeyValueMapperTopologyTest.java:79]: ah, makes sense. Thought this referred to the proto-tests....

File: error-handling-core/build.gradle.kts [hunk 45]
```diff
@@ -0,0 +1,60 @@
+description = "A library for error handling in Kafka Streams."
+
+plugins {
+    id("com.github.davidmc24.gradle.plugin.avro") version "1.2.1"
+}
+
+sourceSets {
+    test {
+        resources {
+            srcDirs("src/test/avro")
```
[the remark this answers is not in the document]
Comment (mkcode92) [reply error-handling-core/build.gradle.kts:10]: To test the avro conversion in ErrorUtilTest

File: error-handling-core/src/main/java/com/bakdata/kafka/ErrorCapturingProcessorContext.java [hunk 46]
```diff
@@ -48,18 +48,4 @@ public <K, V> void forward(final K key, final V value) {
         final ProcessedKeyValue<Object, Object, V> recordWithOldKey = getValue(value);
         super.forward(key, recordWithOldKey);
     }
-
-    @Deprecated
```
[the remark this answers is not in the document]
Comment (mkcode92) [reply error-handling-core/src/main/java/com/bakdata/kafka/ErrorCapturingProcessorContext.java:52]: yes

File: error-handling-proto/build.gradle.kts (hunk 44 above)
Comment (mkcode92) [reply error-handling-proto/build.gradle.kts]: This helped a confused intellij, but the gradle build works also without it, so I removed it again...

File: error-handling-core/build.gradle.kts (hunk 45 above)
[the remark this answers is not in the document]
Comment (mkcode92) [reply error-handling-core/build.gradle.kts]: removed, is automatically handled by the plugin

**Review by philipp94831 — APPROVED** [review]

**Review by philipp94831 — APPROVED** [review]
